### PR TITLE
Feature/composed dashs crud and composition

### PR DESCRIPTION
**Pull request by imenattatra**
Closes https://github.com/metr-systems/backlog/issues/4754
Closes https://github.com/metr-systems/backlog/issues/4755

AI Disclaimer: Assisted by Claude, these PR is mostly crud operations and UI for it, so I generated the code and then reviewed it

## Summary

A new feature called that lets you combine multiple sub-dashboards (coming from se template dashboards) into one dashboard and manage them

- You can create a new composed dashboard from the UI

Screenshot 2026-08-18 at 14 59 26

- You can add sub-dashboards into it from the UI

Screenshot 2026-08-18 at 15 00 09

- You can remove sub-dashboards from it from the UI
- You can change the order of the sub-dashboards in a composed dashboard "edit composition" page

Screenshot 2026-08-18 at 15 00 21

- You can delete a composed dashboard completely
Screenshot 2026-08-18 at 15 00 31

## Coding Strategy

- Started with basic save/read/delete operations for composed dashboards (the CRUD part)
- Added the ability to manage sub-dashboards inside them: adding, removing, and reordering
- Followed existing app patterns on the frontend so it looks and feels consistent with other dashboard features
- Fixed edge cases like preventing duplicate dashboard names and handling auto-increment correctly

## QA

- Unit tests (pytest)
- Manually on staging, you can check it here https://global-dashboard.staging.metr.systems/composed-dashboards , credentials are on lastpass

### CODE REVIEW

File: redash_global/views/composed_dashboards.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        "url_identifier": composed_dashboard.url_identifier,
+        "created_at": composed_dashboard.created_at,
+        "updated_at": composed_dashboard.updated_at,
+    }
+
+
+@login_required
+def composed_dashboards_list():
+    page = int(request.args.get("page", 1))
+    page_size = int(request.args.get("page_size", 25))
```
Comment (berinhard) [inline redash_global/views/composed_dashboards.py]: **Blocker**

We should avoid 500 errors if the querystring gets build like `?page=str-value`. Same for the `page_size`.

File: redash_global/views/composed_dashboards.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    query = ComposedDashboard.query.order_by(ComposedDashboard.created_at.desc())
+
+    total = query.count()
+    composed_dashboards = query.offset((page - 1) * page_size).limit(page_size).all()
```
Comment (berinhard) [inline redash_global/views/composed_dashboards.py:36]: **Blocker**

Considering this `page - 1`, we should also avoid page values lower or equal to 0.

File: redash_global/views/composed_dashboards.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    return jsonify(serialize(composed_dashboard))
+
+
+@login_required
+def composed_dashboard_create():
+    body = request.get_json(silent=True) or {}
+    name = body.get("name")
+    url_identifier = body.get("url_identifier")
+
+    composed_dashboard = ComposedDashboard(name=name, url_identifier=url_identifier)
```
Comment (berinhard) [inline redash_global/views/composed_dashboards.py:60]: **Blocker**

What will happen if the JSON post is something like `{"name": "", "url_identifier": ""}`? I think we should enforce validation on these both these fields being present to prevent errors of partially configured `ComposedDashboard`

File: redash_global/views/composed_dashboards.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    composed_dashboard = ComposedDashboard.query.get_or_404(composed_dashboard_id)
+    entries = composed_dashboard.entries
+    return jsonify([serialize_entry(entry) for entry in entries])
+
+
+@login_required
+def composed_dashboard_entry_create(composed_dashboard_id):
+    ComposedDashboard.query.get_or_404(composed_dashboard_id)
+    body = request.get_json(silent=True) or {}
+    template_dashboard_id = body.get("template_dashboard_id")
```
Comment (berinhard) [inline redash_global/views/composed_dashboards.py:99]: **Blocker**

Same question about input validation here. What does it happens when this information is in they JSON by with empty/invalid data?

File: redash_global/views/composed_dashboards.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    db.session.delete(entry)
+    db.session.commit()
+    return "", 204
+
+
+@login_required
+def composed_dashboard_entries_reorder(composed_dashboard_id):
+    composed_dashboard = ComposedDashboard.query.get_or_404(composed_dashboard_id)
+    body = request.get_json(silent=True) or {}
+    entry_ids = body.get("entry_ids", [])
```
Comment (berinhard) [inline redash_global/views/composed_dashboards.py:134]: **Blocker**

Same about input validaiton. What if the paylod comes with an `entry_ids=42` or even an string, which is valid iterable and will work within the for-loop.

**Review by berinhard — CHANGES_REQUESTED** [review]
The code looks good to me, but I feel we are missing input cleansing in the endpoints that receives data.

**Review by thiagogds — APPROVED** [review]

File: redash_global/views/composed_dashboards.py (hunk 3 above)
Comment (imenattatra) [reply redash_global/views/composed_dashboards.py:60]: `name` and `url_identifier` are being required from the frontend side and this is only consumed by the frontend, I understand the worry but would it still be a blocker in that context?

File: redash_global/views/composed_dashboards.py (hunk 4 above)
Comment (imenattatra) [reply redash_global/views/composed_dashboards.py:99]: the value for template_dashboard_id when creating a composed dashboard entry is selected from a dropdown with available valid values and this is only consumed there

File: redash_global/views/composed_dashboards.py (hunk 5 above)
Comment (imenattatra) [reply redash_global/views/composed_dashboards.py:134]: this value is never directly given by the user , it should always come valid from the frontend

File: redash_global/views/composed_dashboards.py (hunk 1 above)
Comment (imenattatra) [reply redash_global/views/composed_dashboards.py]: Thank you, please find it updated here https://github.com/metr-systems/redash/pull/85/changes/ee32a4e63cfbfc1a76863c17e9c7211c788a5514

File: redash_global/views/composed_dashboards.py (hunk 2 above)
Comment (imenattatra) [reply redash_global/views/composed_dashboards.py:36]: Thank you, this is handled also in this commit https://github.com/metr-systems/redash/commit/ee32a4e63cfbfc1a76863c17e9c7211c788a5514

**Comment by imenattatra** [comment]
> The code looks good to me, but I feel we are missing input cleansing in the endpoints that receives data.

@berinhard 

For the validation concerns (name, url_identifier, template_dashboard_id, entry_ids), these are already validated on the frontend and only used by our frontend code. Given that we have a [planned Pydantic integration](https://github.com/metr-systems/backlog/issues/4981) , I think it makes sense to do it at this point I believe.
Would you be open to reconsidering these as blockers? Knowing that it is needed to merge that ASAP because it will make the composed dashboard deployment tests much easier and faster. 

Thank you for your review!

**Review by berinhard — APPROVED** [review]
As we have the pydantic related issue, I'm fine on moving the blockers here to become expected part of the work from the issue.

**Comment by imenattatra** [comment]
> As we have the pydantic related issue, I'm fine on moving the blockers here to become expected part of the work from the issue.

Thank you !
